### DISCUSSION TITLE
Standardize how models can be reused to rollout for new datasets (e.g. new countries/year)

### DISCUSSION THREAD

**Discussion by butchtm** [thinkingmachines/unicef-ai4d-research-bank#121] [Ideas]
Hi @AnthonyMockler 
Just wanted to get your feedback on how we can standardize the implementation of  #5.

One way to implement this is to recommend that model providers create a docker image that encapsulates an environment that can run model inference and to provide additional instructions on how they can input their own data and extract the results.  As an example, this is how the [Radiant ML Hub ](https://mlhub.earth/models) serve up their models.

IMO, this is a better alternative than asking them to setup a virtual env (for python) and something else if its R or Julia, as it can apply to almost any model.

I think this is probably doable for the poverty mapping project, since it can fetch most of the datasets it needs (via geowrangler modules), assuming they are available for that country/year.
 
[See this example](https://mlhub.earth/model/model-cv4a-crop-detection-v1)
![Screenshot 2023-04-20 110937](https://user-images.githubusercontent.com/82074733/233248770-77014439-357b-4c48-822b-ebe661ffd3d2.png)

**Comment by AnthonyMockler** [comment]
I like the idea (for us). Not sure how we can make this easy / straightforward for outside contributors, but I like it.

**Reply by butchtm** [reply]
We can try it for [poverty mapping](https://github.com/thinkingmachines/unicef-ai4d-poverty-mapping) -- build a docker image to host the python environment and its dependencies as well as the notebooks.  

The poverty mapping project is ideal for this because:
  - it already uses notebooks for all its processing (training dataset prep, model training/eval, grid generation/ country rollout)
  - except for the dhs datasets which is only needed for model training (providing the target labels), all the the other datasets it needs for rolling out EXISTING models for new countries/new years are public datasets and the code to access and cache these datasets are already included in the code (mostly via geowrangler modules). 

To run the notebooks, we can:
  - use  [papermill](https://github.com/nteract/papermill) to run the notebooks via the CLI (on the docker container).
      we can parameterize the country/year so that model users can  just pass them in when rolling the model on new countries/new years
      as well the mount the directories where users can cache the results.

 The tasks related to the poverty mapping project would be the following:
 * Add a docker image (Dockerfile)
 * Add some minor updates to the [notebooks to accept papermill parameters](https://github.com/nteract/papermill#parameterizing-a-notebook)
 * Add instructions on how to run the docker image to pass year/country parameters and mount the cache and output directories (to hold the downloaded public datasets as well as results.
 * Test the whole process
 * Provide links to the instructions in  the model catalog item related links section to guide potential model users
